Docs: add obstore tutorial (#527)
microsoft/PlanetaryComputerDataCatalog: docs/overview/obstore.md

# Reading Planetary Computer data with obstore

> [obstore](https://developmentseed.org/obstore/) is a Python library for reading and writing cloud object stores (Azure Blob, Amazon S3, Google Cloud Storage) directly through their native APIs. Using obstore, SAS tokens refresh automatically, async I/O is built in, and the same store you build for reading bytes can be handed to higher-level libraries like [async-geotiff](https://github.com/developmentseed/async-geotiff), [Lonboard](https://developmentseed.org/lonboard/), and [zarr-python](https://zarr.dev/) without re-authenticating.

Reviewer: > directly through their native APIs

I think this is a bit misleading. I think most users would understand "native" to mean "the raw underlying API specific to each cloud storage provider". That's _not_ what obstore does; if a user wants to use the Azure API directly, they'll use [`azure.storage`](https://pypi.org/project/azure-storage/) directly.

Obstore presents _one, unified, abstracted API_ that is _the same_ across Azure, S3, and GCS. That's the selling point.

> [obstore](https://developmentseed.org/obstore/) is a Python library for reading and writing cloud object stores (Azure Blob, Amazon S3, Google Cloud Storage) directly through their native APIs. Using obstore, SAS tokens refresh automatically, async I/O is built in, and the same store you build for reading bytes can be handed to higher-level libraries like [async-geotiff](https://github.com/developmentseed/async-geotiff), [Lonboard](https://developmentseed.org/lonboard/), and [zarr-python](https://zarr.dev/) without re-authenticating.

Reviewer: ```suggestion
[obstore](https://developmentseed.org/obstore/) is a Python library for reading and writing cloud object stores (Azure Blob, Amazon S3, Google Cloud Storage) directly through their native APIs. Using obstore, Planetary Computer SAS tokens refresh automatically, async I/O is built in, and the same store you build for reading bytes can be handed to higher-level libraries like [async-geotiff](https://github.com/developmentseed/async-geotiff), [Lonboard](https://developmentseed.org/lonboard/), and [zarr-python](https://zarr.dev/) without re-authenticating.
```
Separately, it's good to specify that the built-in SAS token refresh support is connected to Planetary Computer.

A companion notebook walks through every step end-to-end with live timings. [Open in Planetary Computer Hub](https://pccompute.westeurope.cloudapp.azure.com/compute/hub/user-redirect/git-pull?repo=https://github.com/microsoft/PlanetaryComputerExamples&urlpath=lab/tree/PlanetaryComputerExamples/quickstarts/obstore.ipynb&branch=main)

## Install obstore

> obstore works in any Python project. To get started, install obstore alongside `pystac-client` (for searching the Planetary Computer's STAC API) and the HTTP libraries that power its credential providers:

Reviewer: ```suggestion
obstore works in any Python project. To get started, install obstore alongside `pystac-client` (for searching the Planetary Computer's STAC API) and the HTTP libraries that power its [credential providers](https://developmentseed.org/obstore/latest/authentication/#credential-providers):
```

```bash
uv add obstore pystac-client requests aiohttp aiohttp_retry
```

`requests` powers the sync credential provider; `aiohttp` and `aiohttp_retry` power the async one. Install both unless you know you only need one path.

## Connect to a Planetary Computer asset

The most common starting point is a STAC asset returned from a search. obstore's `PlanetaryComputerCredentialProvider` reads the asset's blob URL and handles SAS token acquisition and refresh for you.

1. Open the Planetary Computer STAC catalog and pick a scene to work with.

   ```python
   import pystac_client
   from obstore.auth.planetary_computer import PlanetaryComputerCredentialProvider

   catalog = pystac_client.Client.open(
       "https://planetarycomputer.microsoft.com/api/stac/v1"
   )
   item = next(catalog.search(collections=["naip"], max_items=1).items())
   asset = item.assets["image"]
   ```

2. Build a credential provider from the asset.

   ```python
   provider = PlanetaryComputerCredentialProvider.from_asset(asset)

>    ```

Reviewer: This makes me realize that `from_asset` is a bit annoying if you want to work with a collection instead of an item. 

I see that the [NAIP Collection JSON](https://planetarycomputer.microsoft.com/api/stac/v1/collections/naip) defines
```json
"msft:storage_account": "naipeuwest"
```
so we could potentially have a `from_collection` constructor too. 

Or maybe `from_asset` should really be renamed to `from_stac`, and support both `Item` and `Collection`? Thoughts?

> 3. Build a store using that provider. The store is your reusable connection to that asset.

Reviewer: It's important to note that the store doesn't just connect to _one asset_; it provides the auth to access _anything in that bucket_ (or I guess "container" in Azure terminology) (except as mentioned below, the `prefix` on the store is currently mounted to this specific file)

   ```python
   from obstore.store import AzureStore

   store = AzureStore(credential_provider=provider)
   ```

## Read bytes from the store

Once you have a working store, obstore exposes three read operations that map directly to native Azure Blob API calls.

1. **Read a byte range.** Useful when you only need part of the file. For example, the first ~16 KB of a Cloud Optimized GeoTIFF. 

   ```python
   import obstore

>    header = obstore.get_range(store, "", start=0, end=16384)

Reviewer: I'd prefer pointing users to the method API rather than the functional API.

I.e. 
```py
store.get_range(..., start=0, end=16384)
```
also it is confusing for end users to pass in `""` here... this works because the "prefix" on the store is pointing to the specific file already. It might be better to have the prefix pointing at the top level bucket directory? Or, we have two workflows "opening the bucket at the root" and "opening with an item prefix (which uses `from_asset`)

   ```

2. **Read multiple byte ranges in a single request.** Cuts round-trip latency when you need several non-contiguous slices of the same file (e.g. multiple COG tiles).

   ```python

>    ranges = obstore.get_ranges(

Reviewer: ditto use `store.get_ranges`

       store, "", starts=[0, 65536], ends=[16384, 81920]
   )
   ```

3. **Read the entire file.** Avoid this for large rasters. Range reads and async (below) exist to avoid this scenario.

   ```python
   buf = obstore.get(store, "").bytes()

>    ```

Reviewer: 1. Ditto on using `store.get` over `obstore.get`

2. It's useful to note that the result of `get` is an _iterator_, so you don't _have_ to collect it into a single buffer if you have a use case that supports iteration over a file's contents.

## Run reads in parallel

> For multi-file workloads like building a mosaic or fetching all bands across all scenes in an AOI, running reads in parallel is faster. obstore exposes async equivalents of every read function (`get_async`, `get_range_async`, etc.) that you can compose with `asyncio.gather`.

Reviewer: ```suggestion
For multi-file workloads like building a mosaic or fetching all bands across all scenes in an AOI, making concurrent requests is faster. obstore exposes async equivalents of every read function (`get_async`, `get_range_async`, etc.) that you can compose with `asyncio.gather`.
```
(technically, parallelism and asynchronous/concurrency are two different things)

Async needs its own credential provider class, `PlanetaryComputerAsyncCredentialProvider`, backed by `aiohttp` instead of `requests`. Same `from_asset()` signature.

```python
import asyncio
from obstore.auth.planetary_computer import PlanetaryComputerAsyncCredentialProvider

async_provider = PlanetaryComputerAsyncCredentialProvider.from_asset(asset)
async_store = AzureStore(credential_provider=async_provider)

async def fetch(start, end):
    return await obstore.get_range_async(async_store, "", start=start, end=end)

> results = await asyncio.gather(*[fetch(i * 4096, (i + 1) * 4096) for i in range(8)])

Reviewer: This is a bad example, because it's making several independent requests for different parts of a file. 

For this use case we should be pointing users towards `store.get_ranges_async`, because under the hood that will combine adjacent ranges into a single network request.

So for example, this example makes independent requests for 0-4096, 4096-8192, etc. But `get_ranges_async` would automatically make just a single request under the hood for `0-32768`, instead of 8 concurrent requests, and that would be a lot faster.

```

This is typically 3–5× faster in practice.

## List objects across a container

To enumerate objects under a prefix ("show me every NAIP scene in Montana in 2023"), build a fresh provider against the container URL instead.

```python
container_provider = PlanetaryComputerCredentialProvider(
    "https://naipeuwest.blob.core.windows.net/naip/"
)
container_store = AzureStore(
    account_name="naipeuwest",
    container_name="naip",
    credential_provider=container_provider,
)

for batch in obstore.list(container_store, prefix="v002/mt/2023/"):
    for entry in batch:
        print(entry["path"], entry["size"])
```

## Hand the store to other libraries

> Any library that accepts an [obspec](https://github.com/developmentseed/obspec)-compatible store reads through your authenticated connection without re-doing auth. Open the same NAIP scene as a Cloud Optimized GeoTIFF using async-geotiff:

Reviewer: ```suggestion
Any library that accepts an [obspec](https://github.com/developmentseed/obspec)-compatible store reads through your authenticated connection without re-doing auth. Open the same NAIP scene as a Cloud Optimized GeoTIFF using [async-geotiff](https://github.com/developmentseed/async-geotiff):
```
Doesn't hurt to have another link

## Migrate from `planetary_computer.sign()` + fsspec

If you're updating an existing project, here's the side-by-side. The old pattern:

```python
import planetary_computer
import fsspec

signed = planetary_computer.sign(asset.href)
with fsspec.open(signed) as f:
    data = f.read()
```

The obstore equivalent:

```python
from obstore.auth.planetary_computer import PlanetaryComputerCredentialProvider
from obstore.store import AzureStore
import obstore

provider = PlanetaryComputerCredentialProvider.from_asset(asset)
store = AzureStore(credential_provider=provider)
data = obstore.get(store, "").bytes()
```

obstore handles re-signing on expiry, talks to Azure's native blob API instead of routing through HTTP via fsspec, and exposes async I/O for parallel reads — all without changing your auth code per request.

## Use the same code against other clouds

obstore implements the [obspec](https://github.com/developmentseed/obspec) protocol, so the same read and write calls work against S3 or GCS. Any library built on obspec inherits this portability automatically.

```python
from obstore.store import S3Store

s3_store = S3Store(bucket="my-bucket", region="us-west-2")

> buf = obstore.get(s3_store, "path/to/object").bytes()

Reviewer: Actually this doesn't work... `obstore.get` _won't_ work against the obspec protocol... The obspec protocol is defined in terms of the _methods_ on the class. That's part of why I want to nudge people to use `store.get` instead of `obstore.get`
